presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -94,395 +117,640 @@
-    private final Table materializedView;
-    private final Query materializedViewQuery;
     private final LogicalRowExpressions logicalRowExpressions;
-
-    private MaterializedViewInfo materializedViewInfo;
-    private Optional<Identifier> removablePrefix = Optional.empty();
-    private Optional<Set<Expression>> expressionsInGroupBy = Optional.empty();
+    private final Map<QualifiedObjectName, List<ConnectorMaterializedViewDefinition>> baseTableToMaterializedViewMap;
 
     public MaterializedViewQueryOptimizer(
             Metadata metadata,
             Session session,
             SqlParser sqlParser,
             AccessControl accessControl,
             RowExpressionDomainTranslator domainTranslator,
-            Table materializedView,
-            Query materializedViewQuery)
+            Map<QualifiedObjectName, List<ConnectorMaterializedViewDefinition>> baseTableToMaterializedViewMap)
     {
         this.metadata = requireNonNull(metadata, "metadata is null");
         this.session = requireNonNull(session, "session is null");
         this.sqlParser = requireNonNull(sqlParser, "sql parser is null");
         this.accessControl = requireNonNull(accessControl, "access control is null");
         this.domainTranslator = requireNonNull(domainTranslator, "row expression domain translator is null");
-        this.materializedView = requireNonNull(materializedView, "materialized view is null");
-        this.materializedViewQuery = requireNonNull(materializedViewQuery, "materialized view query is null");
+        this.baseTableToMaterializedViewMap = requireNonNull(baseTableToMaterializedViewMap, "base table to materialized view map is null");
         FunctionAndTypeManager functionAndTypeManager = metadata.getFunctionAndTypeManager();
         logicalRowExpressions = new LogicalRowExpressions(
                 new RowExpressionDeterminismEvaluator(functionAndTypeManager),
                 new FunctionResolution(functionAndTypeManager),
                 functionAndTypeManager);
     }
 
-    public Node rewrite(Node node)
+    @Override
+    protected Node visitQuerySpecification(QuerySpecification node, Void context)
     {
-        // TODO: Implement ways to handle non-optimizable query without throw/catch. https://github.com/prestodb/presto/issues/16541
-        try {
-            MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
-            materializedViewInformationExtractor.process(materializedViewQuery);
-            materializedViewInfo = materializedViewInformationExtractor.getMaterializedViewInfo();
-            return process(node);
-        }
-        catch (Exception ex) {
-            logger.warn("Failed to rewrite query with materialized view with following exception: %s", ex.getMessage());
+        if (!node.getFrom().isPresent()) {
+            log.debug("Visited QuerySpecification had no from field: \n%s", node);
```

Comment:
1. Log it as warning, by default debug messages are not logged.
2. Logging the query node can be lots of information unnecessary informaiton in the log.
3. Change the message to "Query with no from clause are not optimized by materialized view"